compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -3425,3 +3431,59 @@ fn add_lld_args(
         }
     }
 }
+
+// gold has been deprecated with binutils 2.44
+// and is known to behave incorrectly around Rust programs.
+// There have been reports of being unable to bootstrap with gold:
+// https://github.com/rust-lang/rust/issues/139425
+// Additionally, gold miscompiles SHF_GNU_RETAIN sections, which are
+// emitted with `#[used(linker)]`.
+fn warn_if_linked_with_gold(sess: &Session, path: &Path) -> Result<(), Box<dyn std::error::Error>> {
+    use object::elf;
+    use object::read::elf::FileHeader;
+
+    let data = std::fs::read(path)?;
+
+    let was_linked_with_gold = if sess.target.pointer_width == 64 {
+        let elf = elf::FileHeader64::<object::Endianness>::parse(&*data)?;
+        elf_has_gold_version_note(elf, &data)
+    } else if sess.target.pointer_width == 32 {
+        let elf = elf::FileHeader32::<object::Endianness>::parse(&*data)?;
+        elf_has_gold_version_note(elf, &data)
+    } else {
+        return Ok(());
+    };
+
+    match was_linked_with_gold {
+        Ok(true) => {
+            let mut warn = sess
+                .dcx()
+                .struct_warn("the gold linker is deprecated and has known bugs with Rust");
+            warn.help("consider using LLD or ld from GNU binutils instead");
+            warn.emit();
+        }
+        Ok(false) => {}
+        Err(err) => info!(?err, "Error while trying to detect whether binary was linked with gold"),
+    }
+    Ok(())
+}
+
+fn elf_has_gold_version_note(
+    elf: &impl object::read::elf::FileHeader,
+    data: &[u8],
+) -> Result<bool, Box<dyn std::error::Error>> {
+    use object::read::elf::SectionHeader;
```

Comment:
Maybe put `elf_has_gold_version_note` inside `warn_if_linked_with_gold`? That way the use item of `warn_if_linked_with_gold` is in-scope for this function too. This function shouldn't be used outside of `warn_if_linked_with_gold` anyway.